sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -152,22 +159,116 @@ case class TransformWithStateExec(
   override protected def doExecute(): RDD[InternalRow] = {
     metrics // force lazy init at driver
 
-    child.execute().mapPartitionsWithStateStore[InternalRow](
-      getStateInfo,
-      schemaForKeyRow,
-      schemaForValueRow,
-      numColsPrefixKey = 0,
-      session.sqlContext.sessionState,
-      Some(session.sqlContext.streams.stateStoreCoordinator),
-      useColumnFamilies = true
-    ) {
-      case (store: StateStore, singleIterator: Iterator[InternalRow]) =>
-        val processorHandle = new StatefulProcessorHandleImpl(store, getStateInfo.queryRunId)
-        assert(processorHandle.getHandleState == StatefulProcessorHandleState.CREATED)
-        statefulProcessor.init(processorHandle, outputMode)
-        processorHandle.setHandleState(StatefulProcessorHandleState.INITIALIZED)
-        val result = processDataWithPartition(singleIterator, store, processorHandle)
-        result
+    // If the query is running in batch mode, we need to create a new StateStore and instantiate
+    // a temp directory on the executors in mapPartitions
+    if (isStreaming) {
+      child.execute().mapPartitionsWithStateStore[InternalRow](
+        getStateInfo,
+        schemaForKeyRow,
+        schemaForValueRow,
+        numColsPrefixKey = 0,
+        session.sqlContext.sessionState,
+        Some(session.sqlContext.streams.stateStoreCoordinator),
+        useColumnFamilies = true
+      ) {
+        case (store: StateStore, singleIterator: Iterator[InternalRow]) =>
+          processData(store, singleIterator)
+      }
+    } else {
+      child.execute().mapPartitionsWithIndex[InternalRow](
+        (i, iter) => {
+          val providerId = new StateStoreProviderId(
+            StateStoreId(Utils.createTempDir().getAbsolutePath,
+              i, 0), getStateInfo.queryRunId)
+
+          // Create StateStoreProvider for this partition
+          val stateStoreProvider = StateStoreProvider.createAndInit(
+            providerId,
+            schemaForKeyRow,
+            schemaForValueRow,
+            numColsPrefixKey = 0,
+            useColumnFamilies = true,
+            storeConf = StateStoreConf(getDefaultStateStoreSQLConf),
```

Comment:
Instead of trying to copy the conf here, could we do one of the following:

- Create the RocksDB provider explicitly and then call init. Something like this

```
    val provider = new RocksDBStateStoreProvider()
    val sqlConf = new SQLConf()
    val storeConf = new StateStoreConf(sqlConf)

    provider.init(
      storeId, keySchema, valueSchema, 0, useColumnFamilies = true,
      storeConf, new Configuration)
```

- Pass the SQL conf setting and then call createAndInit
```
    val sqlConf = new SQLConf()
   sqlConf.setConfString("<provider class>", "<RocksDBStateStoreProvider>")
    val storeConf = new StateStoreConf(sqlConf)

    StateStoreProvider.createAndInit(
      storeId, keySchema, valueSchema, 0, useColumnFamilies = true,
      storeConf, new Configuration)
```

Reply:
That way we can avoid changes to `StateStoreConf.scala`

Reply:
Right, refactored